iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/testing_error_handler.hpp

```diff
@@ -38,52 +42,68 @@ namespace testing
 {
 
 /// @brief Defines the test reaction of dynamic error handling.
-class TestErrorHandler : public iox::er::ErrorHandlerInterface
+class TestingErrorHandler : public iox::er::ErrorHandlerInterface
 {
   public:
-    TestErrorHandler();
-
-    ~TestErrorHandler() override = default;
-    TestErrorHandler(const TestErrorHandler&) = delete;
-    TestErrorHandler(TestErrorHandler&&) = delete;
-    TestErrorHandler& operator=(const TestErrorHandler&) = delete;
-    TestErrorHandler operator=(TestErrorHandler&&) = delete;
+    TestingErrorHandler() noexcept;
+    ~TestingErrorHandler() noexcept override = default;
+    TestingErrorHandler(const TestingErrorHandler&) noexcept = delete;
+    TestingErrorHandler(TestingErrorHandler&&) noexcept = delete;
+    TestingErrorHandler& operator=(const TestingErrorHandler&) noexcept = delete;
+    TestingErrorHandler operator=(TestingErrorHandler&&) noexcept = delete;
+
+    /// @brief Initialized the error handler. This should be called in the main function of the test binary
+    /// @code
+    /// #include "iceoryx_hoofs/testing/error_reporting/testing_error_handler.hpp"
+    ///
+    /// #include "test.hpp"
+    ///
+    /// int main(int argc, char* argv[])
+    /// {
+    ///     ::testing::InitGoogleTest(&argc, argv);
+    ///
+    ///     iox::testing::ErrorHandler::init();
+    ///
+    ///     return RUN_ALL_TESTS();
+    /// }
+    /// @endcode
+    static void init() noexcept;
 
     /// @brief Defines the reaction on panic.
-    void onPanic() override;
+    void onPanic() noexcept override;
```

Comment:
I think all these handlers (`onXYZ`) should not be `noexcept`. In theory the handler code should be flexible to throw if it desires. I do not see a reason to restrict it.

Generally I am not a fan of `noexcept` in generic code. This includes ode that calls a (potentially polymorphic) function that is decided on runtime and could throw for all we know. In these cases the compiler can never know whether it will throw.

But most importantly the code should not restrict any framework configuring a handler to throw. We can argue that in this particular override in the test handler does not throw, but with C++17 `noexcept` is part of the signature. I have not checked, but specifying the derived version as `noexcept` would require the base interface function to be `noexcept` as well. This precludes any overrides that may throw, which is undesirable.

Reply:
@elBoberido It is still noexcept. Is there a strong reason why? Maybe I was wrong above or did not consider something. Or it was simply forgotten. I mean it apparently compiles (C++17?) so it is probably fine. I still am not a huge fan of littering everything with `noexcept` in generic code

The specialization here is not generic and we know that it does not throw, so it is probably fine. I cannot point to it but I think I miss something important here. Oh well, if so, we will find out later.

Reply:
Are you sure ? `onPanic` has no `noexcept` anymore and it is the only override. All the other methods do not throw exceptions and there are now hooks to inject them.